libmscore/spanner.cpp

```diff
@@ -489,14 +489,37 @@ void Spanner::insertTimeUnmanaged(const Fraction& fromTick, const Fraction& len)
 
 //---------------------------------------------------------
 //   scanElements
-//    used in palettes
 //---------------------------------------------------------
 
 void Spanner::scanElements(void* data, void (* func)(void*, Element*), bool all)
 {
-    Q_UNUSED(all);
-    for (SpannerSegment* seg : segments) {
-        seg->scanElements(data, func, true);
+    bool v = true;
+    if (anchor() == Spanner::Anchor::SEGMENT || anchor() == Spanner::Anchor::CHORD) {
+        Element* se = startElement();
+        Element* ee = endElement();
+        bool v1 = true;
+        if (se && se->isChordRest()) {
+            ChordRest* cr = toChordRest(se);
+            Measure* m    = cr->measure();
+            v1            = m->visible(cr->staffIdx());
+        }
+        bool v2 = true;
+        if (!v1 && ee && ee->isChordRest()) {
+            ChordRest* cr = toChordRest(ee);
+            Measure* m    = cr->measure();
+            v2            = m->visible(cr->staffIdx());
+        }
+        v = v1 || v2;       // hide spanner if both chords are hidden
+    }
+    for (ScoreElement* el : *this) {
+        if (el->isSpannerSegment()) {
+            SpannerSegment* ss = toSpannerSegment(el);
+            if (all || (score()->staff(staffIdx())->show() && v) || isVolta()) {
+                ss->scanElements(data, func, all);
+            }
+        } else {
+            el->scanElements(data, func, all);
+        }
```

Comment:
Please explain why all this is necessary now when it wasn't before. What has changed?

Assuming it is necessary, this is a neater way to do it:

```c++
static bool isVisibleCR(Element* e)
{
    if (!e || !e->isChordRest()) {
        return true; // assume visible
    }
    ChordRest* cr = toChordRest(e);
    return cr->measure()->visible(cr->staffIdx());
}

void Spanner::scanElements(void* data, void (* func)(void*, Element*), bool all)
{
    for (ScoreElement* el : *this) {
        if (el->isSpannerSegment()
            && !all
            && !isVolta()
            && !score()->staff(staffIdx())->show()
            && !isVisibleCR(startElement())
            && !isVisibleCR(endElement())) {
            continue; // don't scan hidden spanner segment
        }
    }
    el->scanElements(data, func, all);
}
```

Reply:
This was actually taken from https://github.com/musescore/MuseScore/blob/master/libmscore/system.cpp#L1196-L1224. According to the comment "// hide spanner if both chords are hidden" it seems it should be necessary, but I'll just check it.

Reply:
Now I've moved this back to the system class and I've avoided changing the existing lines.

Reply:
I actually think it would be better in `SpannerSegment::scanElements`. Where possible, each class should handle its own special cases rather than relying on parents to know about them. (So far tuplet number seems to be the only case where it's not possible and it has to be done by the parent.) This is probably the most elegant way to do it:

```c++
static bool isVisibleCR(Element* e)
{
    if (!e || !e->isChordRest()) {
        return true; // assume visible
    }
    ChordRest* cr = toChordRest(e);
    return cr->measure()->visible(cr->staffIdx());
}

bool Spanner::eitherEndVisible()
{
    if (isVolta()) {
        return true;
    }
    if (!score()->staff(staffIdx())->show()) {
        return false;
    }
    return isVisibleCR(startElement()) || isVisibleCR(endElement());
}

void SpannerSegment::scanElements(void* data, void (* func)(void*, Element*), bool all)
{
    if(all || spanner()->eitherEndVisible()) {
        func(data, this);
    }
}
```

Reply:
Okay, moved to spanner.cpp using these functions.